### PR TITLE
Stac 13274: custom device support

### PR DESCRIPTION
**Pull request by vishalxebia**
### Step 3: Did you add / update tests for your changes in the right area?
- [ ] Unit/Component test
- [ ] Integration test	

### Step 4: I'm confident that everything is properly tested:
I got a PO / QA Approval by:
- [ ] Name

### Step 5: Can we ship this feature to production?
- [ ] Yes, I'm proud of my work. Ship it! :ship:

### CODE REVIEW

File: dynatrace/stackstate_checks/dynatrace/dynatrace.py [hunk 1]
```diff
@@ -19,13 +19,17 @@
 RELATIVE_TIME = 'hour'
 ENVIRONMENT = 'production'
 DOMAIN = 'dynatrace'
+CUSTOM_DEVICE_DEFAULT_RELATIVE_TIME = '3h'
```
Comment (vishalxebia) [inline dynatrace/stackstate_checks/dynatrace/dynatrace.py:22]: fix 1 h.

File: dynatrace/stackstate_checks/dynatrace/dynatrace.py [hunk 2]
```diff
@@ -158,16 +237,26 @@ def _collect_topology(self, response, component_type, instance_info):
         :param instance_info: instance configuration
         :return: create the component on stackstate API
         """
+        # since custom device response returns entities
```
Comment (zandrewitte) [inline dynatrace/stackstate_checks/dynatrace/dynatrace.py:240]: Move this to `process_custom_device_topology` then you don't have to pollute `_collect_topology` too much

File: dynatrace/stackstate_checks/dynatrace/dynatrace.py [hunk 3]
```diff
@@ -284,21 +384,22 @@ def _get_labels(self, dynatrace_component):
                 labels.append("managementZones:%s" % zone.get("name"))
         if dynatrace_component.entityId:
             labels.append(dynatrace_component.entityId)
-        if dynatrace_component.monitoringState:
+        if dynatrace_component.get('monitoringState'):
```
Comment (zandrewitte) [inline dynatrace/stackstate_checks/dynatrace/dynatrace.py:387]: Isn't this a schematics model anymore?

File: dynatrace/stackstate_checks/dynatrace/dynatrace.py [hunk 4]
```diff
@@ -284,21 +384,22 @@ def _get_labels(self, dynatrace_component):
...
-            if technologies.get('type'):
-                tech_label += technologies['type']
-            if technologies.get('edition'):
-                tech_label += ":%s" % technologies['edition']
-            if technologies.get('version'):
-                tech_label += ":%s" % technologies['version']
-            labels.append(tech_label)
+        if dynatrace_component.get('softwareTechnologies'):
+            for technologies in dynatrace_component.softwareTechnologies:
+                tech_label = ''
```
Comment (zandrewitte) [inline dynatrace/stackstate_checks/dynatrace/dynatrace.py:395]: You can use something like: 
```
':'.join(filter(None, [technologies.get('type'), techno...]))
```
to clean this up a bit

File: dynatrace/stackstate_checks/dynatrace/dynatrace.py (hunk 3 above)
Comment (vishalxebia) [reply dynatrace/stackstate_checks/dynatrace/dynatrace.py:387]: yeah interestingly not. if I do access some properties which are not part of the model then it fails.

File: dynatrace/stackstate_checks/dynatrace/dynatrace.py [hunk 5]
```diff
@@ -134,22 +150,85 @@ def check(self, instance_info):
...
+    @staticmethod
+    def get_custom_device_params(instance_info, next_page_key=None):
+        """
+        Process the default parameters needed for custom device
+        @param
+        instance_info: Instance configuration schema
+        @return
+        Returns the parameter for custom device entities
+        """
+        if not next_page_key:
```
[the remark this answers is not in the document]
Comment (vishalxebia) [reply dynatrace/stackstate_checks/dynatrace/dynatrace.py:162]: Done.

File: dynatrace/stackstate_checks/dynatrace/dynatrace.py [hunk 6]
```diff
@@ -134,22 +150,85 @@ def check(self, instance_info):
...
+        """
+        # wanted to override the params for custom-device as it is completely different from smartscape
+        params = self.get_custom_device_params(instance_info)
+        response = self._get_dynatrace_json_response(instance_info, endpoint, params)
+        self._collect_topology(response, component_type, instance_info)
+        next_page_key = response.get('nextPageKey')
+        while next_page_key:
+            params = self.get_custom_device_params(instance_info, next_page_key)
+            response = self._get_dynatrace_json_response(instance_info, endpoint, params)
+            self._collect_topology(response, 'custom-device', instance_info)
```
[the remark this answers is not in the document]
Comment (vishalxebia) [reply dynatrace/stackstate_checks/dynatrace/dynatrace.py:190]: Done.

File: dynatrace/stackstate_checks/dynatrace/dynatrace.py (hunk 2 above)
Comment (vishalxebia) [reply dynatrace/stackstate_checks/dynatrace/dynatrace.py:240]: done.

File: dynatrace/stackstate_checks/dynatrace/dynatrace.py [hunk 7]
```diff
@@ -134,22 +150,85 @@ def check(self, instance_info):
...
+        component_type: Type of the component
+        @return
+        None
+        """
+        # wanted to override the params for custom-device as it is completely different from smartscape
+        params = self.get_custom_device_params(instance_info)
+        response = self._get_dynatrace_json_response(instance_info, endpoint, params)
+        self._collect_topology(response, component_type, instance_info)
+        next_page_key = response.get('nextPageKey')
+        while next_page_key:
```
[the remark this answers is not in the document]
Comment (vishalxebia) [reply dynatrace/stackstate_checks/dynatrace/dynatrace.py:187]: But I have to do looping here until `next_page_key` becomes null or empty.

File: dynatrace/stackstate_checks/dynatrace/dynatrace.py (hunk 4 above)
Comment (vishalxebia) [reply dynatrace/stackstate_checks/dynatrace/dynatrace.py:395]: done.

**Review by zandrewitte — APPROVED** [review]

**Pull request merged**